# Project 3

In this project, you will perform a logistic regression on the admissions data we've been working with in projects 1 and 2.

In [293]:
%matplotlib inline
import matplotlib.pyplot as plt
import pandas as pd
import statsmodels.api as sm
import pylab as pl
import numpy as np


In [294]:
df_raw = pd.read_csv("../assets/admissions.csv")
df = df_raw.dropna() 
print df.head()

   admit    gre   gpa  prestige
0      0  380.0  3.61       3.0
1      1  660.0  3.67       3.0
2      1  800.0  4.00       1.0
3      1  640.0  3.19       4.0
4      0  520.0  2.93       4.0


## Part 1. Frequency Tables

#### 1. Let's create a frequency table of our variables

In [295]:
# frequency table for prestige and whether or not someone was admitted
print pd.crosstab(df['admit'], df['prestige'], rownames=['admit'])

prestige  1.0  2.0  3.0  4.0
admit                       
0          28   95   93   55
1          33   53   28   12


## Part 2. Return of dummy variables

#### 2.1 Create class or dummy variables for prestige 

In [296]:
dummy_ranks = pd.get_dummies(df['prestige'], prefix='prestige')
print dummy_ranks.head()

   prestige_1.0  prestige_2.0  prestige_3.0  prestige_4.0
0           0.0           0.0           1.0           0.0
1           0.0           0.0           1.0           0.0
2           1.0           0.0           0.0           0.0
3           0.0           0.0           0.0           1.0
4           0.0           0.0           0.0           1.0


#### 2.2 When modeling our class variables, how many do we need? 



Answer: Three

## Part 3. Hand calculating odds ratios

Develop your intuition about expected outcomes by hand calculating odds ratios.

In [297]:
cols_to_keep = ['admit', 'gre', 'gpa']
handCalc = df[cols_to_keep].join(dummy_ranks.ix[:, 'prestige':])
print handCalc.head()

   admit    gre   gpa  prestige_1.0  prestige_2.0  prestige_3.0  prestige_4.0
0      0  380.0  3.61           0.0           0.0           1.0           0.0
1      1  660.0  3.67           0.0           0.0           1.0           0.0
2      1  800.0  4.00           1.0           0.0           0.0           0.0
3      1  640.0  3.19           0.0           0.0           0.0           1.0
4      0  520.0  2.93           0.0           0.0           0.0           1.0


In [298]:
df.head()

admit    gre   gpa  prestige
0      0  380.0  3.61       3.0
1      1  660.0  3.67       3.0
2      1  800.0  4.00       1.0
3      1  640.0  3.19       4.0
4      0  520.0  2.93       4.0

In [350]:
# crosstab prestige admission 
cols_to_keep = ['admit', 'gre', 'gpa']
handCalc2 = df[cols_to_keep].join(dummy_ranks.ix[:, :-3])
print handCalc2.head()
print('\n')
# frequency table cutting prestige and whether or not someone was admitted
print pd.crosstab(df['admit'], df['prestige'], rownames=['admit'])

   admit    gre   gpa  prestige_1
0      0  380.0  3.61         1.0
1      1  660.0  3.67         0.0
2      1  800.0  4.00         0.0
3      1  640.0  3.19         0.0
4      0  520.0  2.93         1.0


prestige  1.0  2.0  3.0  4.0
admit                       
0          28   95   93   55
1          33   53   28   12


#### 3.1 Use the cross tab above to calculate the odds of being admitted to grad school if you attended a #1 ranked college

In [372]:
k = pd.crosstab(df.admit, df.prestige, rownames=['admit']).iloc[:, 0]
print k

round(k.loc[1] / float(k.loc[0]), 2)

admit
0    28
1    33
Name: 1.0, dtype: int64


1.18

#### 3.2 Now calculate the odds of admission if you did not attend a #1 ranked college

In [358]:
k = pd.crosstab(df.admit, df.prestige, rownames=['admit']).iloc[:, 1:]
k

prestige  2.0  3.0  4.0
admit                  
0          95   93   55
1          53   28   12

#### 3.3 Calculate the odds ratio

In [334]:
round(sum(k.loc[1]) / float(sum(k.loc[0])), 2)

0.38

In [302]:
print np.exp(result.params)

gre             1.002221
gpa             2.180027
prestige_2.0    0.506548
prestige_3.0    0.262192
prestige_4.0    0.211525
intercept       0.020716
dtype: float64


#### 3.4 Write this finding in a sentance: 

Answer: If you attended a #1 ranked college, you are (1.18 / .38 = 3.1) a little more than 3 times more likely to be admitted.

#### 3.5 Print the cross tab for prestige_4

In [378]:
cols_to_keep = ['admit', 'gre', 'gpa']
handCalc4 = df[cols_to_keep].join(dummy_ranks.ix[:, -1:])
print handCalc4.head()

   admit    gre   gpa  prestige_4
0      0  380.0  3.61         0.0
1      1  660.0  3.67         0.0
2      1  800.0  4.00         0.0
3      1  640.0  3.19         1.0
4      0  520.0  2.93         0.0


#### 3.6 Calculate the OR 

In [379]:
k = pd.crosstab(df.admit, df.prestige, rownames=['admit']).iloc[:, 3]
print k

round(k.loc[1] / float(k.loc[0]), 2)

admit
0    55
1    12
Name: 4.0, dtype: int64


0.22

In [376]:
k = pd.crosstab(df.admit, df.prestige, rownames=['admit']).iloc[:, 3:]
k

prestige  4.0
admit        
0          55
1          12

In [382]:
k2 = pd.crosstab(df.admit, df.prestige, rownames=['admit']).iloc[:, :3]
k2

prestige  1.0  2.0  3.0
admit                  
0          28   95   93
1          33   53   28

In [383]:
round(sum(k2.loc[1]) / float(sum(k2.loc[0])), 2)

0.53

#### 3.7 Write this finding in a sentence

Answer: If you attended a #4 ranked college, you are (.22 / .53 = .42) less than half the time to be admitted.

## Part 4. Analysis

In [337]:
df.head()

admit    gre   gpa  prestige
0      0  380.0  3.61       3.0
1      1  660.0  3.67       3.0
2      1  800.0  4.00       1.0
3      1  640.0  3.19       4.0
4      0  520.0  2.93       4.0

In [338]:
# create a clean data frame for the regression
cols_to_keep = ['admit', 'gre', 'gpa']
data = df[cols_to_keep].join(dummy_ranks.ix[:, 'prestige_1': 'prestige_3'])
data.head()

admit    gre   gpa  prestige_1  prestige_2  prestige_3
0      0  380.0  3.61         1.0         0.0         0.0
1      1  660.0  3.67         0.0         1.0         0.0
2      1  800.0  4.00         0.0         0.0         1.0
3      1  640.0  3.19         0.0         0.0         0.0
4      0  520.0  2.93         1.0         0.0         0.0

We're going to add a constant term for our Logistic Regression. The statsmodels function we're going to be using requires that intercepts/constants are specified explicitly.

In [347]:
# manually add the intercept
data['intercept'] = 1.0
data.head()

admit    gre   gpa  prestige_1  prestige_2  prestige_3  intercept
0      0  380.0  3.61         1.0         0.0         0.0        1.0
1      1  660.0  3.67         0.0         1.0         0.0        1.0
2      1  800.0  4.00         0.0         0.0         1.0        1.0
3      1  640.0  3.19         0.0         0.0         0.0        1.0
4      0  520.0  2.93         1.0         0.0         0.0        1.0

#### 4.1 Set the covariates to a variable called train_cols

In [339]:
train_cols = data.columns[1:]
# Index([gre, gpa, prestige_2, prestige_3, prestige_4], dtype=object)

logit = sm.Logit(data['admit'], data[train_cols])

#### 4.2 Fit the model

In [340]:
result = logit.fit()

Optimization terminated successfully.
         Current function value: 0.624206
         Iterations 5


#### 4.3 Print the summary results

In [341]:
print result.summary()

                           Logit Regression Results                           
Dep. Variable:                  admit   No. Observations:                  397
Model:                          Logit   Df Residuals:                      392
Method:                           MLE   Df Model:                            4
Date:                Mon, 15 Aug 2016   Pseudo R-squ.:                0.001077
Time:                        16:55:13   Log-Likelihood:                -247.81
converged:                       True   LL-Null:                       -248.08
                                        LLR p-value:                    0.9701
                 coef    std err          z      P>|z|      [95.0% Conf. Int.]
------------------------------------------------------------------------------
gre            0.0015      0.001      1.493      0.135        -0.000     0.003
gpa           -0.4696      0.181     -2.590      0.010        -0.825    -0.114
prestige_1    -0.3731      0.311     -1.200      0.2

#### 4.4 Calculate the odds ratios of the coeffiencents and their 95% CI intervals

hint 1: np.exp(X)

hint 2: conf['OR'] = params
        
           conf.columns = ['2.5%', '97.5%', 'OR']

In [309]:
params = result.params
conf = result.conf_int()
conf['OR'] = params
conf.columns = ['2.5%', '97.5%', 'OR']
print np.exp(conf)

                  2.5%     97.5%        OR
gre           1.000074  1.004372  1.002221
gpa           1.136120  4.183113  2.180027
prestige_2.0  0.272168  0.942767  0.506548
prestige_3.0  0.133377  0.515419  0.262192
prestige_4.0  0.093329  0.479411  0.211525
intercept     0.002207  0.194440  0.020716


#### 4.5 Interpret the OR of Prestige_2

Answer: If you attended a #2 ranked college, you are half (.50) the times likely to be admitted. 

#### 4.6 Interpret the OR of GPA

Answer: You are twice (2.18) as likely to be admitted with each unit increase in GPA.

## Part 5: Predicted probablities


As a way of evaluating our classifier, we're going to recreate the dataset with every logical combination of input values. This will allow us to see how the predicted probability of admission increases/decreases across different variables. First we're going to generate the combinations using a helper function called cartesian (above).

We're going to use np.linspace to create a range of values for "gre" and "gpa". This creates a range of linearly spaced values from a specified min and maximum value--in our case just the min/max observed values.

In [311]:
def cartesian(arrays, out=None):
    """
    Generate a cartesian product of input arrays.
    Parameters
    ----------
    arrays : list of array-like
        1-D arrays to form the cartesian product of.
    out : ndarray
        Array to place the cartesian product in.
    Returns
    -------
    out : ndarray
        2-D array of shape (M, len(arrays)) containing cartesian products
        formed of input arrays.
    Examples
    --------
    >>> cartesian(([1, 2, 3], [4, 5], [6, 7]))
    array([[1, 4, 6],
           [1, 4, 7],
           [1, 5, 6],
           [1, 5, 7],
           [2, 4, 6],
           [2, 4, 7],
           [2, 5, 6],
           [2, 5, 7],
           [3, 4, 6],
           [3, 4, 7],
           [3, 5, 6],
           [3, 5, 7]])
    """

    arrays = [np.asarray(x) for x in arrays]
    dtype = arrays[0].dtype

    n = np.prod([x.size for x in arrays])
    if out is None:
        out = np.zeros([n, len(arrays)], dtype=dtype)

    m = n / arrays[0].size
    out[:,0] = np.repeat(arrays[0], m)
    if arrays[1:]:
        cartesian(arrays[1:], out=out[0:m,1:])
        for j in xrange(1, arrays[0].size):
            out[j*m:(j+1)*m,1:] = out[0:m,1:]
    return out

In [312]:
# instead of generating all possible values of GRE and GPA, we're going
# to use an evenly spaced range of 10 values from the min to the max 
gres = np.linspace(data['gre'].min(), data['gre'].max(), 10)
print gres
# array([ 220.        ,  284.44444444,  348.88888889,  413.33333333,
#         477.77777778,  542.22222222,  606.66666667,  671.11111111,
#         735.55555556,  800.        ])
gpas = np.linspace(data['gpa'].min(), data['gpa'].max(), 10)
print gpas
# array([ 2.26      ,  2.45333333,  2.64666667,  2.84      ,  3.03333333,
#         3.22666667,  3.42      ,  3.61333333,  3.80666667,  4.        ])


# enumerate all possibilities
combos = pd.DataFrame(cartesian([gres, gpas, [1, 2, 3, 4], [1.]]))

[ 220.          284.44444444  348.88888889  413.33333333  477.77777778
  542.22222222  606.66666667  671.11111111  735.55555556  800.        ]
[ 2.26        2.45333333  2.64666667  2.84        3.03333333  3.22666667
  3.42        3.61333333  3.80666667  4.        ]


#### 5.1 Recreate the dummy variables

In [313]:
# recreate the dummy variables
combos.columns = ['gre', 'gpa', 'prestige', 'intercept']
dummy_ranks = pd.get_dummies(combos['prestige'], prefix='prestige')
dummy_ranks.columns = ['prestige_1', 'prestige_2', 'prestige_3', 'prestige_4']

# keep only what we need for making predictions
cols_to_keep = ['gre', 'gpa', 'prestige', 'intercept']
combos = combos[cols_to_keep].join(dummy_ranks.ix[:, 'prestige_2':])

#### 5.2 Make predictions on the enumerated dataset

In [314]:
# make predictions on the enumerated dataset
combos['admit_pred'] = result.predict(combos[train_cols])

KeyError: "Index([u'prestige_2.0', u'prestige_3.0', u'prestige_4.0'], dtype='object') not in index"

#### 5.3 Interpret findings for the last 4 observations

Answer: 

## Bonus

Plot the probability of being admitted into graduate school, stratified by GPA and GRE score.